File: leetcode-algorithms/002. Add Two Numbers/solution.cpp
#include <iostream>
using namespace std;

struct ListNode {
     int val;
     ListNode *next;
     ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    	ListNode dummy(-1);
    	int carry = 0;
    	ListNode *prev = &dummy;
    	for(ListNode *pa = l1,*pb = l2;pa!=nullptr || pb!=nullptr; 
    		pa = pa==nullptr?nullptr:pa->next,
    		pb = pb==nullptr?nullptr:pb->next,
    		prev = prev->next)
    	{
    		const int ai = pa==nullptr?0:pa->val;
    		const int bi = pb==nullptr?0:pb->val;
    		const int value = (ai+bi+carry)%10;
    		carry = (ai+bi+carry)/10;
    		prev->next = new ListNode(value);
    	}
    	if (carry>0)
    		prev->next = new ListNode(carry);
    	return dummy.next;
    }
};



int main()
{
	Solution s;
	ListNode *l1 = new ListNode(2);
	ListNode *l2 = new ListNode(5);
	l1->next = l2;
	ListNode *l3 = new ListNode(3);
	l2->next = l3;
	ListNode *res = s.addTwoNumbers(l1,l1);
	for(ListNode *p=res;p!=nullptr;p=p->next)
		cout<<p->val<<endl;


}

File: leetcode-algorithms/002. Add Two Numbers/solution.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        dummy = cur = ListNode(0)
        carry = 0
       	while l1 or l2 or carry:
       		if l1:
       			carry += l1.val
       			l1 = l1.next
       		if l2:
       			carry += l2.val
       			l2 = l2.next
       		carry,val = divmod(carry,10)
       		cur.next = ListNode(val)
       		cur = cur.next
       	return dummy.next

if __name__ == '__main__':
	l1 = ListNode(2)
	l2 = ListNode(3)
	l1.next = l2
	l3 = ListNode(4)
	l2.next = l3
	s = Solution()
	res = s.addTwoNumbers(l1,l1)
	while res:
		print(res.val)
		res = res.next